Upgrade Patternfly to version 5.4

ferhoyos opened the pull request:
### Describe the change

Upgrade Patternfly to version 5.4

### Steps to test the PR

Verify that Kiali works as expected and CI tests pass

### Automation testing

N/A

### Issue reference

Closes #7674

ferhoyos commented:
@jshaughn there are differences in the dagre layout of the patternfly graph:

Traffic graph:

5.2.0
![image](https://github.com/user-attachments/assets/69827dd6-e243-4df8-9d30-9136ccfc5765)

5.4.0
![image](https://github.com/user-attachments/assets/980f05a7-8d4a-4738-a573-35b6bf8c973f)


Mesh graph:

5.2.0
![image](https://github.com/user-attachments/assets/697eaff8-baba-4b04-8a3b-ed4c7dab177a)

5.4.0
![image](https://github.com/user-attachments/assets/ba0f88fe-6ecc-4613-ab9b-66827c787c71)


The edges are very long in PF 5.4.0

CI tests fail because typeahead component is not working correctly. I have to check if it is a bug in 5.4.0 or we need to adapt the component to 5.4 version

ferhoyos commented:
Thanks @jshaughn for your dagre suggestions. Now the PF graphs look fine:

![image](https://github.com/user-attachments/assets/db0f9ea5-55bc-44a7-a898-6ee48b98e44d)

![image](https://github.com/user-attachments/assets/f6b9da3b-44fe-459d-afa0-2b26d70b65bd)

jshaughn commented:
Problem: This may be a regression:  Close side-panel correctly resizes the graph, but Open side-panel does not.

jshaughn commented:
Problem: Dark Mode
- Badge background too light
- TCP edge maybe a little too dark

![image](https://github.com/user-attachments/assets/e0ae366b-7b24-41f3-858b-bce5bd20f22d)

jshaughn commented:
Problem: The non-Dagre layouts don't support groups.  So, we need to do something to prevent their use when groups are active (cluster and/or namespace groups enabled in Display menu, and/or Versioned App graph).  Discussion needed...

ferhoyos commented:
@jshaughn collapse side-panel issue is fixed, just need to wait a bit before the graph resize (increase the timeout to 250ms).

Also, I have modified the edge and badge colors to adapt them to dark mode:

Light mode:

![image](https://github.com/user-attachments/assets/e34611a1-d5bb-4967-85b3-d9b14257ee50)

Dark mode:

![image](https://github.com/user-attachments/assets/38855f02-8039-49d5-b10e-e6813cfcfc5e)

ferhoyos commented:
Regarding the non-dagre layout issue, we could consider hiding the non-dagre layout toolbar options whenever there is a group in the graph

```
hidden: graphData?.elements?.nodes?.filter(node => node.data.isBox).length > 0
```

jshaughn commented:
> Regarding the non-dagre layout issue, we could consider hiding the non-dagre layout toolbar options whenever there is a group in the graph

This is a good idea for when the graph type is "versioned app" because the app boxing can not be stopped.  But for other graph types those layouts may be useful, as long as namespace and cluster boxing is off.  We may want to do something a little hacky and for those layouts temporarily ignore those settings, even in enabled.

By the way, maybe you missed https://github.com/ferhoyos/kiali/pull/4?

ferhoyos commented:
Yes, I missed https://github.com/ferhoyos/kiali/pull/4, sorry. I did not receive any notification when the PR was created (as configured by default), so I modified the watch options for my personal repositories to avoid this issue in the future.

ferhoyos commented:
@jshaughn I have disabled groups for non-dagre layouts. 

Dagre layout:

![image](https://github.com/user-attachments/assets/1f623ef3-e7f6-4e9e-b16f-45ac6cdb61c2)

Grid layout:

![image](https://github.com/user-attachments/assets/d054658e-a096-49ac-9b5e-4a638a6ffbab)

I am not sure if we should indicate this limitation in the graph for these layouts.

jshaughn commented:
One more thing I noticed, we need to see if we can improve the labeling to use PF badges as opposed to parentheses.  I remember updating Cytoscape, but not sure why I didn't update PF at the same time.

Cy:
![image](https://github.com/user-attachments/assets/53556dcb-c905-4fe1-99c0-974c634ab7c4)

Pft:
![image](https://github.com/user-attachments/assets/cc1a655f-4fbe-4d32-ad93-5fdd7a321d46)

ferhoyos commented:
> One more thing I noticed, we need to see if we can improve the labeling to use PF badges as opposed to parentheses. I remember updating Cytoscape, but not sure why I didn't update PF at the same time.

We cannot use PF badges instead of parentheses because Patternfly topology only allows string type as secondary label: 

https://github.com/patternfly/react-topology/blob/main/packages/module/src/components/nodes/labels/NodeLabel.tsx#L27

PF Badges can only be displayed in the primary label (e.g., the last row of nodes in https://topology-v5.surge.sh/styles-nav-link/labels-nav-link)

ferhoyos commented on frontend/src/pages/GraphPF/GraphPF.tsx:
```diff
@@ -393,14 +393,24 @@ const TopologyContent: React.FC<{
-          newNode = addGroup(nd as NodeData);
+
+        // Non-dagre layouts do not support groups
+        if (layoutName === LayoutName.Dagre) {
```
This idea looks good to me. I have added `supportsGroups(layoutName)` function in `GraphUtils.ts`. We may need to check grouping layouts in more places, so it's better to have the functionality centralized in one location.

ferhoyos commented:
I have modified the tooltips in PF graph legend to show boxing and non-boxing layouts:

![image](https://github.com/user-attachments/assets/483b2242-3d72-490f-ab38-5bc6e5e55893)


Also, I have added extra info about grouping layout in the help text of `Namespace Boxes` and `Cluster Boxes` display options:

![image](https://github.com/user-attachments/assets/c3e1ad0a-4b42-4c3e-a2bd-3f057a7d369b)

jshaughn commented:
> We cannot use PF badges instead of parentheses because Patternfly topology only allows string type as secondary label:

Well that explains why I didn't do it :-)   There are still just so many little pain points in the PFT variant of the graph.  After the upgrade we need to see if we can really improve a few of these things, I'm still quite afraid users will see this change as a step backwards.  I'm not sure what to point to in the new graph as a real improvement.

The updated info text, and tooltip content, will hopefully be useful.  Thanks for adding it.

jshaughn reviewed (approved):
I think everything related to the upgrade has been addressed, and some additional pain points as well.  Looks good!
